PhysicsTools/NanoAOD/python/taus_cff.py

```diff
@@ -30,23 +35,25 @@
     )
 
 ##################### Tables for final output and docs ##########################
-def _tauIdWPMask(pattern, choices, doc=""):
-    return Var(" + ".join(["%d * tauID('%s')" % (pow(2,i), pattern % c) for (i,c) in enumerate(choices)]), "uint8", 
-               doc=doc+": bitmask "+", ".join(["%d = %s" % (pow(2,i),c) for (i,c) in enumerate(choices)]))
-def _tauId2WPMask(pattern,doc):
-    return _tauIdWPMask(pattern,choices=("Loose","Tight"),doc=doc)
-def _tauId3WPMask(pattern,doc):
-    return _tauIdWPMask(pattern,choices=("Loose","Medium","Tight"),doc=doc)
-def _tauId4WPMask(pattern,doc):
-    return _tauIdWPMask(pattern, choices=("VLoose", "Loose", "Medium", "Tight"), doc=doc)
-def _tauId5WPMask(pattern,doc):
-    return _tauIdWPMask(pattern,choices=("VLoose","Loose","Medium","Tight","VTight"),doc=doc)
-def _tauId6WPMask(pattern,doc):
-    return _tauIdWPMask(pattern,choices=("VLoose","Loose","Medium","Tight","VTight","VVTight"),doc=doc)
-def _tauId7WPMask(pattern,doc):
-    return _tauIdWPMask(pattern,choices=("VVLoose","VLoose","Loose","Medium","Tight","VTight","VVTight"),doc=doc)
-def _tauId8WPMask(pattern,doc):
-    return _tauIdWPMask(pattern,choices=("VVVLoose","VVLoose","VLoose","Loose","Medium","Tight","VTight","VVTight"),doc=doc)
+def _tauIdWPMask(pattern, choices, doc="", from_raw=False, wp_thrs=None):
+    if from_raw:
+        assert wp_thrs is not None, "wp_thrs argument in _tauIdWPMask() is None, expect it to be dict-like"
+        vs_type = pattern.split('VS')[-1]
+        if vs_type not in ['e', 'mu', 'jet']:
+            raise ValueError("in WP from raw computation, vs_type should be in [e,mu,jet], got %s" % vs_type)
+        
+        var_definition = []
+        for wp_name in choices:
+            if not isinstance(wp_thrs[vs_type][wp_name], float):
+                raise TypeError("Threshold for vs_type=%s and WP=%s is not a float number." % (vs_type, wp_name))
+            wp_definition = "(tauID('%s')>%.4f)" % (pattern % wp_name, wp_thrs[vs_type][wp_name])
+            var_definition.append(wp_definition)
+        var_definition = " + ".join(var_definition)
+    else:
+        var_definition = " + ".join(["tauID('%s')" % (pattern % c) for c in choices])
+    
+    doc = doc + ": "+", ".join(["%d = %s" % (i,c) for (i,c) in enumerate(choices, start=1)])
+    return Var(var_definition, "uint8", doc=doc)
```

Comment:
I would keep storing WPs as bits instead of using different integers for this. There are at least two arguments for it:
* it is backward compatible solution (used also for other leptons?);
* it is generally better design (one can better control size of used number).

Reply:
I am more in favour of having WP stored as numbers. Here are some arguments in favour:
- having bit per WP is redundant, i.e. if tau passes a tighter WP it will always pass a looser WP.
- Other objects that store WP in nanoAOD use numbering notation instead of bit masks (e.g. `Electron_cutBased`, `Muon_miniIsoId`, `Muon_mvaId`, `Photon_cutBased`)
- it is more user-friendly and intuitive for newcomers (at least based on my experience with students)
- the nano-creation code becomes simpler (e.g. no need for specialised `_tauIdWPMask`)

Reply:
sorry, small correction: it looks like `Photon_cutBased` is a bitmask, so there is no unified choice for other objects...

Reply:
We can use numbers if you think that it is better choice than setting bits - I tend to agree that is simpler for non-experienced users. However, the change should be clearly communicated to users (including our validation and SF teams) to avoid confusions. BTW, numbers are also partly redundant as our WPs are redundant, i.e. always all WPs looser than a given WP are passed if the WP is passed.

Reply:
> Anyway, I can try to investigate next week, but do you know when the group is expecting to have the code in CMSSW to rerun their ntuples?

Thank you! They were about to start the production the last week, but since we didn't have any recipe at that time, they moved forward without v2p5. It would be great to have everything ready to profit from possible feedback from analyses in the future.